File: weird_captcha_gym/shared_runtime/app/mechanics/clockbeat_catacomb.js
(() => {
'use strict';
const D={UP:[0,-1],RIGHT:[1,0],DOWN:[0,1],LEFT:[-1,0],WAIT:[0,0]}, glyph={UP:'↑',RIGHT:'→',DOWN:'↓',LEFT:'←',WAIT:'·'};
let timer, handler;
const equal=(a,b)=>a[0]===b[0]&&a[1]===b[1];
function render(state,h){
 document.body.dataset.mechanic="clockbeat_catacomb";
 clearInterval(timer);if(handler)window.removeEventListener('keydown',handler);
 const b=state.board, mode=state.control_condition?.interaction||'full';
 const s={player:[...b.start],hp:b.health,enemies:structuredClone(b.enemies),beat:0,used:false,won:false};
 let start=null,elapsed=0,events=[],submitted=false,message='Wind the clock when ready.';
 const ph=e=>(s.beat+e.offset)%({crawler:1,hopper:2,lancer:3}[e.kind]);
 const solid=p=>b.walls.some(q=>equal(p,q));
 function act(key){
  if(s.won||s.hp<=0||s.used)return 'unavailable';s.used=true;
  if(key==='WAIT')return 'wait';
  const [dx,dy]=D[key],p=[s.player[0]+dx,s.player[1]+dy];if(solid(p))return 'wall';
  const e=s.enemies.find(e=>e.hp>0&&equal(e.pos,p));
  if(e){if(e.kind==='hopper'&&ph(e)===1)return 'shield';e.hp--;return e.hp===0?'defeated':'hit';}
  s.player=p;s.won=equal(p,b.exit)&&s.enemies.every(e=>e.hp===0);return s.won?'escaped':'move';
 }
 function step(){
  if(s.won||s.hp<=0)return;
  for(const e of s.enemies){
   if(e.hp<=0)continue;const phase=ph(e);
   if(e.kind==='hopper'&&phase===0)continue;if(e.kind==='lancer'&&phase!==1)continue;
   const [x,y]=e.pos,[px,py]=s.player,dist=Math.abs(x-px)+Math.abs(y-py);
   if(e.kind==='lancer'){if(dist===1)s.hp--;continue;}
   const choices=Object.values(D).map((d,i)=>({d,i,dist:Math.abs(x+d[0]-px)+Math.abs(y+d[1]-py)})).sort((a,b)=>a.dist-b.dist||a.i-b.i);
   for(const {d,i,dist:nd} of choices){if(i===4||nd>=dist)continue;const p=[x+d[0],y+d[1]];
    if(solid(p)||s.enemies.some(o=>o!==e&&o.hp>0&&equal(o.pos,p)))continue;
    if(equal(p,s.player))s.hp--;else e.pos=p;break;
   }
  }s.beat++;s.used=false;
 }
 function sync(){
  if(start===null||s.won||s.hp<=0||submitted)return;
  elapsed=Math.min(b.max_beats*b.beat_ms,Math.floor(performance.now()-start));
  while(s.beat<Math.floor(elapsed/b.beat_ms)&&s.hp>0&&!s.won)step();
  if(s.beat>=b.max_beats)message='CLOCK EXHAUSTED · Submit to try a fresh catacomb.';
 }
 h.app.innerHTML=`<section class="cc" tabindex="0"><header><div><small>THE UNDERGROUND HOROLOGICAL SOCIETY</small><h1>Clockbeat Catacomb</h1></div><div class="cc-health"></div></header><main><div class="cc-chamber"><div class="cc-board" style="--n:${b.size}"></div><div class="cc-verdict"></div></div><aside><small>SHARED BEAT</small><div class="cc-pulse"><i></i></div><h2 class="cc-beat">CLOCK AT REST</h2><p class="cc-window">One move or attack while the bar is gold.</p><div class="cc-roster"></div><p class="cc-rule">Move into a sentry to attack. At the end of every beat, sentries act in numbered order. Chasers favor ↑ → ↓ ← on ties; blocked paths do not detour. Clear every sentry, then enter the stair.</p>${mode==='full'?'<div class="cc-keys">← ↑ ↓ →<small>ARROWS / WASD · SPACE TO WAIT</small></div>':`<div class="cc-buttons">${Object.keys(D).map(k=>`<button data-cc-key="${k}" aria-label="${k}">${glyph[k]}</button>`).join('')}</div>`}</aside></main><footer><button class="cc-start">WIND THE CLOCK</button><div class="cc-message"></div><button class="cc-submit">SUBMIT EXPEDITION</button></footer><div class="readout" data-status="idle"></div></section>`;
 const root=h.app.querySelector('.cc'),board=root.querySelector('.cc-board');
 function draw(){
  const active=start!==null&&!s.won&&s.hp>0&&s.beat<b.max_beats;
  root.querySelector('.cc-health').textContent='LANTERN '+ '♥'.repeat(Math.max(0,s.hp))+'♡'.repeat(Math.max(0,b.health-s.hp));
  root.querySelector('.cc-beat').textContent=start===null?'CLOCK AT REST':`BEAT ${s.beat+1} · ${s.used?'PLAYED':elapsed%b.beat_ms<b.open_ms?'ACT NOW':'LISTEN'}`;
  const pulse=root.querySelector('.cc-pulse');pulse.style.setProperty('--progress',`${100*(elapsed%b.beat_ms)/b.beat_ms}%`);pulse.classList.toggle('closed',elapsed%b.beat_ms>=b.open_ms);
  root.querySelector('.cc-message').textContent=message;
  root.querySelector('.cc-roster').innerHTML=s.enemies.map(e=>`<div class="cc-entry ${e.kind}"><b>${e.id} · ${{crawler:'Brass Mite',hopper:'Vault Beetle',lancer:'Chime Thorn'}[e.kind]}</b><span>${e.hp<=0?'DISMANTLED':e.kind==='crawler'?'CHASE → CHASE':e.kind==='hopper'?(ph(e)===0?'REST / OPEN → SHIELD + CHASE':'SHIELD + CHASE → REST / OPEN'):['PREPARE → STRIKE','STRIKE → REST','REST → PREPARE'][ph(e)]}</span><small>${e.kind==='lancer'?'Strikes four adjacent tiles; stays rooted.':e.kind==='hopper'?'Armored during chase. Attack while resting.':'Chases one tile; contact costs one heart.'}</small></div>`).join('');
  let cells='';for(let y=0;y<b.size;y++)for(let x=0;x<b.size;x++){
   const p=[x,y],wall=solid(p),e=s.enemies.find(e=>e.hp>0&&equal(e.pos,p)),player=equal(s.player,p),exit=equal(b.exit,p);
   const danger=s.enemies.some(e=>e.hp>0&&e.kind==='lancer'&&ph(e)===1&&Math.abs(x-e.pos[0])+Math.abs(y-e.pos[1])===1);
   cells+=`<div class="cc-tile ${wall?'wall':''} ${danger&&!wall?'danger':''}">${exit?`<span class="cc-stair ${s.enemies.every(e=>e.hp===0)?'unlocked':''}">≋</span>`:''}${e?`<span class="cc-enemy ${e.kind} ${e.kind==='hopper'&&ph(e)===1?'shield':''}"><b>${e.id}</b><i>${'•'.repeat(e.hp)}</i></span>`:''}${player?'<span class="cc-player"><i></i></span>':''}</div>`;
  }board.innerHTML=cells;
  root.querySelector('.cc-verdict').textContent=s.won?'STAIR OPEN · ESCAPED':s.hp<=0?'LANTERN OUT · FAILED':'';
  root.querySelector('.cc-verdict').classList.toggle('visible',s.won||s.hp<=0);
  root.dataset.beat=String(s.beat);root.dataset.used=String(s.used);root.dataset.ms=String(elapsed);root.dataset.won=String(s.won);
 }
 function input(key,source){
  sync();if(start===null||submitted||s.won||s.hp<=0||s.beat>=b.max_beats)return;
  const outcome=elapsed%b.beat_ms>=b.open_ms?'offbeat':act(key);
  events.push({sequence:events.length+1,ms:elapsed,key,input_source:source,outcome});
  message={offbeat:'OFF BEAT · No move. Sentries still advance.',unavailable:'BEAT ALREADY PLAYED · Wait for the next gold bar.',shield:'ARMORED · Attack blocked.',wall:'WALL · Beat spent.',defeated:'SENTRY DISMANTLED',hit:'HIT · Sentry damaged.',move:'STEP ACCEPTED',wait:'WAIT ACCEPTED',escaped:'ESCAPED · Submit expedition.'}[outcome];draw();
 }
 root.querySelector('.cc-start').onclick=()=>{if(start!==null)return;h.setReadout('', 'idle');start=performance.now();message='The clock is running.';root.querySelector('.cc-start').disabled=true;draw();};
 root.querySelectorAll('[data-cc-key]').forEach(el=>el.onclick=()=>input(el.dataset.ccKey,'direction_buttons'));
 handler=e=>{if(mode!=='full'||e.repeat)return;const key={ArrowUp:'UP',ArrowRight:'RIGHT',ArrowDown:'DOWN',ArrowLeft:'LEFT',w:'UP',d:'RIGHT',s:'DOWN',a:'LEFT',' ':'WAIT'}[e.key];if(key){e.preventDefault();input(key,'keyboard');}};window.addEventListener('keydown',handler);
 root.querySelector('.cc-submit').onclick=async()=>{
  if(submitted)return;sync();submitted=true;
  const payload={mechanic_id:state.mechanic_id,task_id:state.task_id,challenge_id:state.challenge_id,actions:events,final_ms:elapsed,final_state:structuredClone(s)};
  try{const response=await fetch('/result',{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify(payload)});const out=await response.json();
   if(out.passed){root.querySelector('.cc-verdict').textContent='PASS · EXPEDITION VERIFIED';root.querySelector('.cc-verdict').classList.add('visible');h.setReadout('PASS','passed');clearInterval(timer);}
   else if(out.state){await h.render(out.state);h.setReadout('FAIL · Fresh catacomb ready','error');}
   else{submitted=false;message='Submission unavailable. Try again.';draw();}
  }catch(e){submitted=false;message='Submission unavailable. Try again.';draw();}
 };
 timer=setInterval(()=>{sync();draw();},50);draw();root.focus();
}
window.WeirdCaptchaMechanics=window.WeirdCaptchaMechanics||{};window.WeirdCaptchaMechanics.clockbeat_catacomb={rootSelector:'.cc',render};
})();

File: weird_captcha_gym/shared_runtime/server/incubator_graders/clockbeat_catacomb.py
"""Independent beat-gated replay. No client outcome is trusted."""
from copy import deepcopy
import math
MECHANIC_ID = 'clockbeat_catacomb'
DELTAS = {'UP':(0,-1),'RIGHT':(1,0),'DOWN':(0,1),'LEFT':(-1,0),'WAIT':(0,0)}

def initial(b):
    return {'player':b['start'][:], 'hp':b['health'], 'enemies':deepcopy(b['enemies']), 'beat':0, 'used':False, 'won':False}

def phase(e, beat):
    return (beat + e['offset']) % {'crawler':1,'hopper':2,'lancer':3}[e['kind']]

def act(b,s,key):
    if s['won'] or s['hp']<=0 or s['used']: return 'unavailable'
    s['used']=True
    if key=='WAIT': return 'wait'
    dx,dy=DELTAS[key]; p=[s['player'][0]+dx,s['player'][1]+dy]
    if p in b['walls']: return 'wall'
    e=next((e for e in s['enemies'] if e['hp']>0 and e['pos']==p),None)
    if e:
        if e['kind']=='hopper' and phase(e,s['beat'])==1: return 'shield'
        e['hp']-=1
        return 'defeated' if e['hp']==0 else 'hit'
    s['player']=p
    s['won']=p==b['exit'] and all(e['hp']==0 for e in s['enemies'])
    return 'escaped' if s['won'] else 'move'

def advance(b,s):
    if s['won'] or s['hp']<=0:return
    for e in s['enemies']:
        if e['hp']<=0:continue
        ph=phase(e,s['beat'])
        if e['kind']=='hopper' and ph==0:continue
        if e['kind']=='lancer' and ph!=1:continue
        x,y=e['pos']; px,py=s['player']; distance=abs(x-px)+abs(y-py)
        if e['kind']=='lancer':
            # Telegraph is the four adjacent tiles, drawn by the browser.
            if distance==1:s['hp']-=1
            continue
        choices=sorted(enumerate(DELTAS.values()),key=lambda v:(abs(x+v[1][0]-px)+abs(y+v[1][1]-py),v[0]))
        for i,(dx,dy) in choices:
            if i==4:continue
            p=[x+dx,y+dy]
            if abs(p[0]-px)+abs(p[1]-py)>=distance:continue
            if p in b['walls'] or any(o is not e and o['hp']>0 and o['pos']==p for o in s['enemies']):continue
            if p==s['player']:s['hp']-=1
            else:e['pos']=p
            break
    s['beat']+=1;s['used']=False

def grade(payload,ground_truth,public_state):
    def fail(msg):return {'graded':True,'passed':False,'feedback':msg}
    if not all(isinstance(value, dict) for value in (payload, ground_truth, public_state)):
        return fail('invalid document')
    for k in ('mechanic_id','task_id','challenge_id'):
        if not ground_truth.get(k) or payload.get(k)!=ground_truth[k] or public_state.get(k)!=ground_truth[k]:return fail(k+' mismatch')
    if ground_truth['mechanic_id']!=MECHANIC_ID:return fail('wrong mechanic')
    if public_state.get('board')!=ground_truth.get('board') or public_state.get('control_condition')!=ground_truth.get('control_condition'):return fail('contract mismatch')
    b=ground_truth['board'];s=initial(b)
    source='keyboard' if (ground_truth.get('control_condition') or {}).get('interaction','full')=='full' else 'direction_buttons'
    events=payload.get('actions');end=payload.get('final_ms')
    def valid_time(t):return type(t) in (int,float) and math.isfinite(t) and 0<=t<=b['max_beats']*b['beat_ms']
    if not isinstance(events,list) or len(events)>1000 or not valid_time(end):return fail('invalid transcript')
    last=0
    for i,e in enumerate(events):
        if not isinstance(e,dict) or e.get('sequence')!=i+1 or e.get('input_source')!=source or not isinstance(e.get('key'), str) or e['key'] not in DELTAS:return fail('invalid input')
        t=e.get('ms')
        if not valid_time(t) or t<last or t>end:return fail('invalid time')
        if t >= b['max_beats']*b['beat_ms']:return fail('input after clock exhausted')
        if s['won'] or s['hp']<=0:return fail('input after terminal state')
        while s['beat']<int(t//b['beat_ms']) and s['hp']>0:advance(b,s)
        if s['hp']<=0:return fail('input after death')
        outcome='offbeat' if t%b['beat_ms']>=b['open_ms'] else act(b,s,e['key'])
        if e.get('outcome')!=outcome:return fail('action outcome mismatch')
        last=t
    while s['beat']<int(end//b['beat_ms']) and s['hp']>0 and not s['won']:advance(b,s)
    passed=s['won'] and s['hp']>0 and payload.get('final_state')==s
    return {'graded':True,'passed':passed,'feedback':f"{'PASS' if passed else 'FAIL'} · health {s['hp']} · sentries {sum(e['hp']>0 for e in s['enemies'])} · exit {s['won']}"}

import heapq, itertools
def find_route(board,state=None,limit=100000):
    start=deepcopy(state or initial(board));serial=itertools.count()
    def ident(s):return (tuple(s['player']),s['hp'],s['beat']%6,s['used'],tuple((tuple(e['pos']),e['hp']) for e in s['enemies']))
    def estimate(s):
        live=[e for e in s['enemies'] if e['hp']>0];p=s['player']
        return sum(e['hp'] for e in live)+(min(abs(p[0]-e['pos'][0])+abs(p[1]-e['pos'][1]) for e in live) if live else abs(p[0]-board['exit'][0])+abs(p[1]-board['exit'][1]))
    q=[(0,next(serial),0,start,[])];best={ident(start):0}
    while q and len(best)<limit:
        _,_,cost,s,path=heapq.heappop(q)
        if s['won']:return path
        if cost>60:continue
        for key in DELTAS:
            n=deepcopy(s);out=act(board,n,key)
            if out in ('wall','unavailable','shield'):continue
            if not n['won']:advance(board,n)
            if n['hp']<=0:continue
            nc=cost+1+(s['hp']-n['hp'])*12
            identn=ident(n)
            if nc>=best.get(identn,10**9):continue
            best[identn]=nc;heapq.heappush(q,(nc+3*estimate(n),next(serial),nc,n,path+[key]))
    raise RuntimeError('No route found within planner budget')
